NXP S32 introduce support PSI5

congnguyenhuu opened the pull request:
Introduce support Peripheral Sensor Interface (PSI5) driver on S32Z. This driver allows to communication (send, receive) with PSI5 device.
The test result: [PSI5-TESTING-RESULT.pptx](https://github.com/user-attachments/files/17946327/PSI5-TESTING-RESULT.pptx)

RFC: https://github.com/zephyrproject-rtos/zephyr/issues/83982

congnguyenhuu committed:
drivers: introduce support Peripheral Sensor Interface (PSI5) driver

This driver allows to communication (send, receive) with PSI5 device

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
tests/samples: drivers: create test, sample for psi5 driver

Create test, sample for psi5 driver

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

congnguyenhuu committed:
boards: s32z270: enable support psi5

enable support psi5

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

JarmouniA reviewed (commented):
A test is needed as well, samples are not a replacement for tests.

rruuaanng commented on samples/boards/nxp/s32/psi5/src/main.c:
Because this is the smallest sample it provides that can run, I guess its prototype is developed based on nxp soc.

congnguyenhuu replied:
I moved this sample to a generic driver sample.
And also created a test API driver .

congnguyenhuu commented on dts/bindings/psi5/psi5-controller.yaml:
```diff
@@ -0,0 +1,152 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+description: Properties for Peripheral Sensor Interface (PSI5)
```
I updated

rruuaanng replied:
@congnguyenhuu They are redundant and do not need to be prefixed. Putting them in `dts/bindings` already indicates that they are a binding. E.g. other bindings description.
```
Renesas RA ADC node

Static Battery Characteristics

Description of pre-defined file systems.

ADI MAX32 compatible Counter RTC

PL330 DMA Controller
```

congnguyenhuu replied:
I updated

congnguyenhuu commented on samples/boards/nxp/s32/psi5/src/main.c:
```diff
@@ -0,0 +1,48 @@
+
+#include <zephyr/drivers/psi5/psi5.h>
+
+#define PSI5_NODE DT_INST(0, nxp_s32_psi5)
```
I updated

congnguyenhuu commented:
.

congnguyenhuu commented:
Based on the suggestions from the Architecture WG meeting on March 4th, I have updated to the v1 changes. These updates include:
1. Changed `add_rx_callback(dev, channel, serial_cb, data_cb, user_data)` to
`register_callback(dev, channel, struct psi5_rx_callback_configs callback_configs)`
 
With the newly defined structure for RX callback configurations in `psi5.h`
```
struct psi5_rx_callback_config {
	psi5_rx_frame_callback_t callback;
	struct psi5_frame *frame;
	uint32_t max_num_frame;
	void *user_data;
};
 
struct psi5_rx_callback_configs {
	struct psi5_rx_callback_config *serial_frame;
	struct psi5_rx_callback_config *data_frame;
};
```
 
2. RX buffers are allocated by the user to improve throughput:
- Use the API `register_callback` to register the RX callback and allocate a user-defined RX buffer for each callback.
- Changed the callbcak handler:
- Count the number of RX buffer frames and trigger the RX callback when the count reaches the maximum buffer frame limit.
- Reset the count when a RX error occurs or when the count reaches the maximum buffer frame limit.
 
3. Tests and samples have been updated to demonstrate usage with the new changes

congnguyenhuu commented:
New update as Dat's suggestion for [PR](https://github.com/zephyrproject-rtos/zephyr/pull/80117)

New updated PSI5 frame type, separate 2 type serial format 4 bit ID and 8 bit ID
```
enum psi5_frame_type {
	PSI5_SERIAL_FRAME_4_BIT_ID,
	PSI5_SERIAL_FRAME_8_BIT_ID,
	PSI5_DATA_FRAME
};
```

New updated PSI5 status, uses the common status RX frame and RX error all type frames
```
enum psi5_status {
	PSI5_TX_DONE,
	PSI5_TX_ERR,
	PSI5_RX_FRAME,
	PSI5_RX_ERR,
};
```

congnguyenhuu commented:
I have updated to the v2 changes as [the suggestion for SENT](https://github.com/zephyrproject-rtos/zephyr/pull/80117#discussion_r2076726019).
I removed the `enum psi5_status`.
Tx callback still report the `status` with changed type to `int`, `status = 0` means tx done else tx error.
Rx callback just report the `num_frame`,  An error frame is received (number received frames != expected amount), only report non-error frame.

Dat-NguyenDuy commented on dts/bindings/psi5/psi5-controller.yaml:
```diff
@@ -0,0 +1,153 @@
+          - xx-long-frame: XX-Long (V2.0)
+          - non-standard-frame: Non Standard Length
+
+    num-rx-buf:
```
it seems to be this is vendor specific property?

congnguyenhuu replied:
I moved this to vendor S32 binding

Dat-NguyenDuy commented on dts/bindings/psi5/psi5-controller.yaml:
```diff
@@ -0,0 +1,153 @@
+        Selects the receive message bitrate in kbps. This setting determines the speed at
+        which data is received.
+
+    array-slot-duration-us:
```
I think it's better to use child binding for the slot, should we consider to use it instead of the array?

congnguyenhuu replied:
I used a child binding for these config for channel rx slot

congnguyenhuu replied:
I updated to use child binding for these config for channel rx slot

Dat-NguyenDuy commented on dts/bindings/psi5/psi5-controller.yaml:
```diff
@@ -0,0 +1,153 @@
+        falling edge of a synchronization pulse, measured in microseconds (us). This value
+        determines the delay before the decoder starts processing incoming signals again.
+
+    tx-mode:
```
should we use `tx-frame` or the like? to avoid duplicating `frame` within the enum value

Dat-NguyenDuy commented on include/zephyr/drivers/psi5/psi5.h:
```diff
@@ -0,0 +1,247 @@
+ * @param status     PSI5 status.
+ * @param user_data  User data provided when the frame was sent.
+ */
+typedef void (*psi5_tx_callback_t)(const struct device *dev, uint8_t channel, int status,
```
It seems like the definition for a possible value of `status` is missing.

congnguyenhuu replied:
I mentioned the possible values in description

Dat-NguyenDuy commented on drivers/psi5/psi5_nxp_s32.c:
```diff
@@ -0,0 +1,519 @@
+	struct psi5_nxp_s32_channel_data *channel_data = &data->channel_data[channel];
+	int err;
+
+	if (!(config->channel_mask & BIT(channel))) {
```
Please move all these condition checks to `psi5_nxp_s32_send`

congnguyenhuu replied:
I upadted

congnguyenhuu replied:
I updated

Dat-NguyenDuy commented on include/zephyr/drivers/psi5/psi5.h:
```diff
@@ -0,0 +1,247 @@
+ * @retval -EIO if a general transmit error occurred.
+ * @retval -EAGAIN on timeout.
+ */
+__syscall int psi5_send(const struct device *dev, uint8_t channel, const uint64_t data,
```
Any pre-condition before this API is invoked, like channel shall be configured first in sync mode and sync pulse already generated? Pls mention in API prototype

congnguyenhuu replied:
I mentioned  the pre-condition

Dat-NguyenDuy commented on include/zephyr/drivers/psi5/psi5.h:
```diff
@@ -0,0 +1,247 @@
+__syscall int psi5_register_callback(const struct device *dev, uint8_t channel,
+				     struct psi5_rx_callback_configs callback_configs);
+
+static inline int z_impl_psi5_register_callback(const struct device *dev, uint8_t channel,
```
Any pre-condition before this API is invoked? Pls mention in API prototype.

congnguyenhuu replied:
I mentioned  the pre-condition

congnguyenhuu commented on dts/bindings/psi5/psi5-controller.yaml:
```diff
@@ -0,0 +1,153 @@
+        asynchronous mode with only the receive function active. When set to false,
+        the channel operates in synchronous mode with both transmit and receive functions active.
+
+    period-sync-pulse-us:
```
I updated

manuargue commented:
hi @carlescufi , @henrikbrixandersen . Can we remove the arch review label or is there anything else to discuss? Thanks

congnguyenhuu commented on include/zephyr/drivers/psi5/psi5.h:
```diff
@@ -0,0 +1,254 @@
+
+/**
+ * @brief PSI5 Interface
+ * @defgroup psi5_interface PSI5 Interface
```
I updated this

dleach02 commented on include/zephyr/drivers/psi5/psi5.h:
```diff
@@ -0,0 +1,278 @@
+			/** Serial message data */
+			uint16_t data;
+		} serial;
+	};
```
This might vary between architectures but this is talking to an on-chip IP block so different architectures is not an issue. But I generally agree though that the sub structure should put the uint8 rsvd field where it explicitly expects it in the byte stream. It could be before the id, between the id & data, or after as you suggested. 

But having said that, as implemented, it works with the underlying hardware so the alignment likely meets the expectations of the HAL functions and the IP on chip. We would like to get this merged within the scope of the merge window because it also involved a HAL update that has been merged which complicates things. 

If we determine there is a bug here, we can patch it after the feature freeze.

dkalowsk replied:
@dleach02 I understand your concern and desire to get this merged.  And you call out the alternate case I am concerned about.
```diff
		struct {
			/** Serial message ID */
			uint8_t id;
			/** Serial message data */
			uint16_t data;
+			uint8_t rsvd;
		} serial;
```
and
```diff
		struct {
			/** Serial message ID */
			uint8_t id;
+			uint8_t rsvd;
			/** Serial message data */
			uint16_t data;
		} serial;
```
both have different meanings and support/output.  I'm not sure I follow the argument that this is an IP block only so it's correct.  The compiler will still need to provide the data structure implementation to be used.  Running this through CompilerExplorer in a simplified example shows my concern as the second is the case being generated: https://gcc.godbolt.org/z/Prb9b1bjn

manuargue replied:
`serial` and `data` represent mutually exclusive frame formats, meaning the union is used to hold either a serial frame or a data frame, and they are never meant to be read or written together. So assuming that the consumer follow the intended use, there won't be undefined behaviors due to padding, neither aliasing risks.

That said, I noticed this hasn't been properly documented so it opens possibility of misuse. Additionally, using a tagged union could improve code clarity without sacrificing compactness or requiring to model each frame type as a complete separate struct. For example:
```c
struct psi5_frame {
    enum psi5_frame_type type;
    union {
        struct {
            uint32_t payload;
        } data;  /* access this when .type is PSI5_DATA_FRAME */
        struct {
            uint8_t id;
            uint16_t payload;
        } serial;   /* access this when .type is PSI5_SERIAL_FRAME_4_BIT_ID or PSI5_SERIAL_FRAME_8_BIT_ID */
    };
    uint32_t timestamp;
    uint8_t crc;
    uint8_t slot_number;
};
```

dleach02 replied:
@dkalowsk sorry for not being clear. What I meant in my comment about it being an IP block is that this is an IP block where we are introducing a new driver framework. Our S32 platform is the only one in Zephyr that will be enabled for this IP block. It isn't generic code that will be compiled into any additional platforms so the exposure is only the NXP S32 and nothing else in Zephyr. 

But having said that, I was trying to get this PR merged before the feature freeze because we had all the approvals already. In anticipation of this, we merged our HAL PR which now causes us sequencing problems (you can chat with Daniel about this because he understands the larger challenges we have). 

I conceded to you that we did indeed need to address this but suggested that we would address this as a bug fix after the feature was merged. You seemed to not respond to this request and instead continued down the path of the problem you identified. Again, I conceded that point.

Now I'm in a situation where I'm forced to request a variance from the release team and/or TSC to have this merged.

rruuaanng commented on dts/bindings/psi5/psi5-controller.yaml:
```diff
@@ -0,0 +1,141 @@
+    tx-frame:
+      type: string
+      enum:
+        - short-31-1s
```
For string, you should not use implicit.
```suggestion
        - "short-31-1s"
```
(And other.)
